### DISCUSSION TITLE
do-mpc as a framework for open-loop dynamic optimization

### DISCUSSION THREAD

**Discussion by pavaninguva** [do-mpc/do-mpc#307] [Q&A] [answered]
Hi, I have a bit of a conceptual question here. 

If I were to run a do-mpc code for a state-space model (i.e. using `do_mpc.estimator.StateFeedback(model)`) with no parameter uncertainty, it seems reasonable to assume that the generated results for `_x` and `_u` for do-mpc would be equivalent to an open-loop dynamic optimization / trajectory optimization (the literature has so many different terms for this). 

I understand it is possible to run an open-loop simulation (https://www.do-mpc.com/en/latest/example_gallery/DIP.html?highlight=open%20loop#Run-open-loop) where we simply run the model once i.e. `u0 = mpc.make_step(x0)` and simply set the prediction horizon as long as we would like for an open-loop optimziation. 

The reason I mention this is because for a case I am considering, do-mpc (both open and closed-loop) generates highly comparable results to another optimization suite (Gekko) to solve a problem: 

do-mpc results (open-loop):
![Figure_4](https://user-images.githubusercontent.com/41192490/193482155-0dde982a-fd2e-412e-aee3-cebe89b058a7.png)

do-mpc results (closed-loop):
![Figure_3](https://user-images.githubusercontent.com/41192490/193481970-8b658afc-fb3c-4e2e-b0e6-6c46d8ce4284.png)

Gekko results (open-loop):  
![TIV_Opt_Low](https://user-images.githubusercontent.com/41192490/193481981-deb7a380-e28f-4293-966a-950f7e126d6a.png)

So I have the following questions / observations I would be grateful for any help on:

1. Is a state-space MPC run (i..e, closed-loop) with no uncertain parameters and disturbances equivalent to an open-loop optimization?  My intuition tells me this is the case, because even in the DIP example, the closed-loop system follows the same trajectory as that generated from the open-loop solution. 

2. The reason I ask this question is because for a slightly more complex system than the one presented above, it turns out other optimization suites in Python and Julia (e.g., infiniteopt.jl / gekko / pyomo) are unable to reasonably solve for an optimal `u0` when running an open-loop optimization, but do-mpc is able to do so quite nicely (even though everyone is using Ipopt as a backend). 

3, If indeed do-mpc can be used for such a use case, that would be amazing! I would be quite happy to contribute the example above as part of the example gallery if it is of interest to the team.

**Comment by 4flixt** [comment, marked as answer]
Hi @pavaninguva,

apologies for the late reply. Regarding you questions: 

1. Without uncertainty and disturbances, the open-loop prediction and the open-loop trajectory (not updating the control policy from the first solution of the MPC problem) should be almost identical. I say almost, because the simulator uses a different method to integrate the ODE/DAE than the MPC solver. Significant differences could arise for excessively large time-steps for example. However, the finite horizon closed-loop MPC trajectory is not going to be identical to the original open-loop prediction. If you want to learn more about this have a look at Bellmanns principle of optimality. Here, the difference are going to be larger for a short prediction horizon (and might vanish for a long prediction horizon).

2. I am not sure I understand your question entirely. Do you mean the behavior of do-mpc when setting the ``open-loop`` setting to true in [mpc.set_param](https://www.do-mpc.com/en/latest/api/do_mpc.controller.MPC.set_param.html#do_mpc.controller.MPC.set_param)? This refers to something different than what you might expect: It is about multi-stage robust MPC (you can read more about this [here](https://www.do-mpc.com/en/latest/theory_mpc.html#robust-multi-stage-nmpc).

I hope this helps! Best,
4flixt
